File: C_plus_plus/Subsets.cpp
class Solution {
public:
    vector<vector<int> > subsets(vector<int> &S) {
        vector<vector<vector<int>>> dp(S.size()+1);
        dp[0] = vector<vector<int>>(1);
        sort(S.begin(),S.end());
        for(int i = 1 ; i <= S.size() ; ++i){
            dp[i] = dp[i-1];
            for(int j = 0 ; j < dp[i-1].size() ; ++j){
                vector<int> tt = dp[i-1][j];
                tt.push_back(S[i-1]);
                dp[i].push_back(tt);
            }
        }
        return dp[S.size()];
    }
};

File: java/Subsets.java
// Iteration method
// Firstly, sort the array.
// And then, we can find the nth's result by using the n-1th result.
// Iterate it, we come out all the subsets.
// Time and space are both O(2^n).

public class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        result.add(new ArrayList<Integer>());
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; ++i) {
            int size = result.size();
            for (int j = 0; j < size; ++j) {
                List<Integer> list = new ArrayList<Integer>(result.get(j));
                list.add(nums[i]);
                result.add(list);
            }
        }
        return result;
    }
}
